### DISCUSSION TITLE
how to export custom fields in pdf template

### DISCUSSION THREAD

**Discussion by urmel1960** [kimai/kimai#2549] [General]
Hi there
I bought the custom field plugin and tried now to export a custom field created with that plugin into pdf template. 

btw: I am looking for a comprehensive list or advise for the place to look at to get a complete list of the possible variables you can use within pdf templates when exporting

I tried a couple of those - but no success ..

I am desperately looking for the proper syntax - any help is highly appreciated.

Thanks and regards

**Comment by kevinpapst** [comment]
If you run in "prod" environment, you have to clear the cache for template changes, because compiled templates they are cached.

There is no such list, but you can always start by examining the HTML-Print export template:
https://github.com/kevinpapst/kimai2/blob/master/templates/export/renderer/default.html.twig

I haven't tested it but I guess in order to export a custom timesheet field, you need one import:
```{% import "macros/widgets.html.twig" as widgets %}```

and then in the loop where you export the timesheet records:
```
{% set metaField = entry.metaField('mein_feld') %}
{{ widgets.form_type_value(metaField.type, metaField.value, entry) }}
```

**Comment by empx92** [comment]
Same issue there.
The above description works well but how do i check if the field is not empty?

Thank you for your help

**Reply by kevinpapst** [reply]
```
{% if metaField.value == '' %}
...
{% endif %}
```

**Reply by empx92** [reply]
Thank you for the quick Answer. But as it seems that will only work within the loop.
But i need to check specific for the metaField if it is empty for example > if meta.Field('et_material') is empty < then i want to genarate the Table. Because if i understand right otherwise it would generate the table too, if for example 1 out of 4 metaFields are filled out.

In my example i would like to generate a list of Materials and that list should be extra for every Customer/Project

{% if metaField.value == '' %}	{# does not work because not within a loop #}
		debug: kein Eintrag
	{% else %}
        {% for entry in entries %}
				{% block material_list %}
					{% block material_list_header %}
						<h3>Material Liste</h3>
				{% endblock %}
			
			<table class="items">
				<thead>
				<tr>
					<th>{{ 'label.date'|trans }}<br />Besteller</th>
					<th>Material</th>
				</tr>
				</thead>
				<tbody>
		
		
            
            <tr class="{{ cycle(['odd', 'even'], loop.index0) }}">
                <td class="text-nowrap">
                    {% block material_date_begin %}{% if not showDateTimeShort %}{{ entry.begin|date_time }}{% else %}{{ entry.begin|date_short }}{% endif %}{% endblock %}
                    {% if showUserColumn %}<br />{{ entry.user.displayName }}{% endif %}
					
                </td>
                <td>
                    {% block material_description %}
                        <b>{{ entry.project.customer.name }} - {{ entry.project.name }}</b>{% if entry.activity is not null %} - <small>{{ entry.activity.name }} - </small>{% endif %}
						{% if entry.description is not empty %} <small>{{ entry.description|desc2html }}</small> {% endif %}
						{% set metaField = entry.metaField('et_material') %} {# Setze den Metafield Namen #}
                            <br>
                            <i style="font-size:10px;">
							{{ widgets.form_type_value(metaField.type, metaField.value, entry) }}</i>
                    {% endblock %}
                </td>
            </tr>
        </tbody>
    </table>
{% endblock %}
{% endfor %}
{% endif %} {# ende abfrage material #}
{# ENDE MATERIALLISTE #}

Thank you very much!

**Reply by kevinpapst** [reply]
Can you exactly describe your use case. Materials are not part of Kimai and I donÄt understand what you want.

**Reply by empx92** [reply]
I have a custom Field which is a Textarea, which is called et_material. This customfield is used in Timesheets.

So for example if i am on a service order and i needed a lightbulb for the repair i can use the custom field "Material" in the timesheet.
If i have 10 Services a Day with different customers and/or additionally a few larger projects i need to export those customfields with my other times in a seperate table and seperate for every customer.

My standard export now looks like:
Page 1 : Summary of all Projects within the export timeframe (usually one week)
Page 3 - x : Full Listing of Date & Time with Customer - Project and - Description

After the standard export i want to give out the custom field (like the not working example code above), so if within the Timesheet there is a customer with a filled out customField Named "et_material" there should be a Table with the whole Materials which for this customer.

**Reply by kevinpapst** [reply]
Aha, that makes it clearer. 
Then in the "customer loop" and before the "timesheet loop" you do `{% set hasMaterial = false %}` 
and within the "timesheet loop" you add the "if has meta field and meta field is not empty" check and if true 
then `{% set hasMaterial = true %}`
After the "timesheet loop" you check for `{% if hasMaterial %}` and then add another "timesheet/material loop"  which only shows those that have a filled metafield.

**Comment by empx92** [comment]
On The Page
https://www.kimai.org/documentation/invoices.html#twig-templates
you can find some information which variables you could use.
Some of them work for me because a custom field in the section "project"  uses `entry.project.metaField('et_projektort')` for example

**Comment by empx92** [comment]
Did someone manage to find out how to export the Customfields for user*?

so far i know:
"Zeiterfassung" (Timesheet) = {% set yourVariable = entry.metaField('yourVariable') %}
Customer = (didn't need, so i don't know)
Project = {% set yourVariable= entry.project.metaField('yourVariable') %}
Activity = (didn't need, so i don't know)
User = *
Invoice = (didn't need, so i don't know)

* tried following not working codes:
{% set et_verwendung = entry.UserPreferences.metaField('et_verwendung') %}
{% set et_verwendung = entry.User_Preferences.metaField('et_verwendung') %}
{% set et_verwendung = entry.user.metaField('et_verwendung') %}

Thank you in advance for your help

**Reply by kevinpapst** [reply]
You are linking the docs chapter yourself... 

```
{{ entry['entry. user_preference.et_verwendung'] }}
```
<img width="590" alt="Bildschirmfoto 2024-05-12 um 11 49 10" src="https://github.com/kimai/kimai/assets/533162/6ad14e42-f7d0-4806-9cbc-1545ad018fa7">

<img width="660" alt="Bildschirmfoto 2024-05-12 um 11 50 05" src="https://github.com/kimai/kimai/assets/533162/7559c822-c0a0-41ee-adcf-12289fa517f0">

You are trying to use a syntax that I do not support.

**Reply by empx92** [reply]
Ah okay thank you, i was totally fixed at this part: yourVariable = entry.**metaField('yourVariable')**
